src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -845,3 +1014,172 @@ impl<T: fmt::Display + fmt::Debug + DisplayableInTimeline + TimestampManipulatio
+        /// Timestamps and reasons that impose an inclusive upper bound.
+        pub upper: Vec<(Antichain<mz_repr::Timestamp>, Reason)>,
+    }
+
+    impl DisplayableInTimeline for Constraints {
+        fn fmt(&self, timeline: Option<&Timeline>, f: &mut fmt::Formatter) -> fmt::Result {
+            if !self.lower.is_empty() {
+                writeln!(f, "lower:")?;
+                for (ts, reason) in &self.lower {
+                    let ts = ts.iter().map(|t| t.display(timeline)).collect::<Vec<_>>();
+                    writeln!(f, "  {:?}: ({:?})", ts, reason)?;
+                }
+            }
+            if !self.upper.is_empty() {
+                writeln!(f, "upper:")?;
+                for (ts, reason) in &self.upper {
+                    let ts = ts.iter().map(|t| t.display(timeline)).collect::<Vec<_>>();
+                    writeln!(f, "  {:?}: ({:?})", ts, reason)?;
+                }
+            }
+            Ok(())
+        }
+    }
+
+    impl Debug for Constraints {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            self.display(None).fmt(f)?;
+            Ok(())
+        }
+    }
+
+    impl Constraints {
+        /// Remove constraints that are dominated by other constraints.
+        ///
+        /// This removes redundant constraints, without removing constraints
+        /// that are "tight" in the sense that the interval will not improve
+        /// without their removal. For example, two constraints at the same
+        /// time will both be retained, in the interest of full information.
+        pub fn minimize(&mut self) {
+            // Establish the upper bound of lower constraints.
+            let lower_frontier = self.lower_bound();
+            // Retain constraints that intersect `lower_frontier`.
+            self.lower.retain(|(anti, _)| {
+                anti.iter()
+                    .any(|time| lower_frontier.elements().contains(time))
+            });
+
+            // Establish the lower bound of upper constraints.
+            let upper_frontier = self.upper_bound();
+            // Retain constraints that intersect `upper_frontier`.
+            self.upper.retain(|(anti, _)| {
+                anti.iter()
+                    .any(|time| upper_frontier.elements().contains(time))
+            });
+        }
+
+        /// An antichain equal to the least upper bound of lower bounds.
+        pub fn lower_bound(&self) -> Antichain<mz_repr::Timestamp> {
+            let mut lower = Antichain::from_elem(mz_repr::Timestamp::minimum());
+            for (anti, _) in self.lower.iter() {
+                lower = lower.join(anti);
+            }
+            lower
+        }
+        /// An antichain equal to the greatest lower bound of upper bounds.
+        pub fn upper_bound(&self) -> Antichain<mz_repr::Timestamp> {
+            self.upper
+                .iter()
+                .flat_map(|(anti, _)| anti.iter())
+                .cloned()
+                .collect()
```

Comment:
I'm not super familiar with `Antichain`, but I'm assuming that the `FromIterator` impl is doing the bulk of the work to get the "greatest lower bound"?

Reply:
According to frank, this is so

Reply:
Yup. For better or worse, `Antichain` tracks the lower bound of inserted elements, and so blatting everything down and collecting it results in the lower bound of all the elements, which is the greatest lower bound of all the input antichains.